File: src/main/scala/resonant/lib/grid/Node.java
package resonant.lib.grid;

import java.util.AbstractMap;
import java.util.WeakHashMap;

import resonant.api.grid.INode;
import resonant.api.grid.INodeProvider;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraftforge.common.ForgeDirection;

public abstract class Node<P extends INodeProvider, G extends Grid, N> implements INode
{
    public final P parent;

    protected final AbstractMap<N, ForgeDirection> connections = new WeakHashMap<N, ForgeDirection>();

    public G grid = null;

    public Node(P parent)
    {
        this.parent = parent;
    }

    public final G getGrid()
    {
        if (grid == null)
            grid = newGrid();

        return grid;
    }

    protected abstract G newGrid();

    public final void setGrid(G grid)
    {
        this.grid = grid;
    }

    @Override
    public void update(float deltaTime)
    {

    }

    /** TODO: Try inject tile validate and invalidate events so this does not have to be called. This
     * constructs the node. It should be called whenever the connections of the node are updated OR
     * when the node is first initiated and can access its connections. */
    @Override
    public void reconstruct()
    {
        recache();
        getGrid().add(this);
        getGrid().reconstruct();
    }

    @Override
    public void deconstruct()
    {
        synchronized (connections)
        {
            /** Remove self from all connections. */
            for (N connection : connections.keySet())
            {
                if (getGrid().isValidNode(connection))
                {
                    ((Node) connection).getConnections().remove(this);
                }
            }

            getGrid().remove(this);
            getGrid().deconstruct();
        }
    }

    /** Called for a node to recache all its connections. */
    @Override
    public final void recache()
    {
        synchronized (connections)
        {
            doRecache();
        }
    }

    protected void doRecache()
    {

    }

    /** All the connections we are connected to. Note that this may not be thread safe for
     * modification unless synchronized.
     * 
     * @return Returns all the connections in this node. */
    public AbstractMap<N, ForgeDirection> getConnections()
    {
        return connections;
    }

    /** Can this node connect with the source?
     * 
     * @param from - Direction coming from.
     * @param source - Object trying to connect with this node. This should either extend Node or be
     * an object that can interface with the node. */
    public boolean canConnect(ForgeDirection from, Object source)
    {
        return false;
    }

    /** Must be called to load the node's data. */
    public void load(NBTTagCompound nbt)
    {

    }

    /** Must be called to save the node's data. */
    public void save(NBTTagCompound nbt)
    {

    }

    @Override
    public String toString()
    {
        return getClass().getSimpleName() + "[" + hashCode() + ", Connections: " + connections.size() + ", Grid:" + getGrid() + "]";
    }
}

File: src/main/scala/resonant/lib/grid/node/Node.scala
package resonant.lib.grid.node

import net.minecraft.entity.Entity
import net.minecraft.tileentity.TileEntity
import net.minecraft.world.World
import resonant.api.grid.{INode, INodeProvider}
import resonant.lib.transform.vector.{TVectorWorld, VectorWorld}

/**
 * A node is any single part of a grid.
 * @author Calclavia
 */
abstract class Node(var parent: INodeProvider) extends INode with TVectorWorld
{
  override def reconstruct()
  {
  }

  override def deconstruct()
  {
  }

  override def getParent: INodeProvider = parent

  override def x: Double =
  {
    return parent match
    {
      case x: TileEntity => x.xCoord
      case x: Entity => x.posX
      case _ => 0
    }
  }

  override def y: Double =
  {
    return parent match
    {
      case x: TileEntity => x.yCoord
      case x: Entity => x.posY
      case _ => 0
    }
  }

  override def z: Double =
  {
    return parent match
    {
      case x: TileEntity => x.zCoord
      case x: Entity => x.posZ
      case _ => 0
    }
  }

  override def world: World =
  {
    return parent match
    {
      case x: TileEntity => x.getWorldObj
      case x: Entity => x.worldObj
      case _ => null
    }
  }

  override def toString: String = getClass.getSimpleName + "[" + hashCode + "]"
}